### ISSUE TITLE
Any way to handle filters other than with Contains?

### ISSUE THREAD

**Issue by 25747** [Groopit/ra-data-odata-server#5]
Hello - This package is working perfectly for the majority of my needs, but I have a few areas where I will want to do more than a "Contains" type filter. For example, checking if a boolean column is true or false, checking for values greater than zero, date ranges, and so on.

Is there any built in (or other recommended) method to handle these types of filtering cases? Alternatively, do you have any recommendations for my backend team to handle these cases on their side?

**Comment by 25747** [comment]
I looked through the code over the weekend, and today pulled the code down to experiment a bit. I appended some suffixes to my filters so that I could process them in the dataprovider. I made a "filterBuilder" function, and I call that for each filter instead of the original `Contains(${filterName},'${params.filter[filterName]}')` template literal: 

```javascript
//around line 101; getList function in the original repo
for (const filterName in params.filter) {
  o = o.filter(
    filterBuilder(filterName, params.filter[filterName])
    //`Contains(${filterName},'${params.filter[filterName]}')`
  );
} //NOTE:  JS because i'm not a TS guy yet :)
```
here is my filterBuilder function:
```javascript
const filterBuilder = (filterName, filterValue) => {
  const [filterResource, filterOperation, valueType] = filterName.split("_");
  if (valueType === "str") filterValue = `'${filterValue}'`; //if filtering on a String field, make sure it has single quotes around value
  if (filterOperation) {
    //if the filter operation exists, build filter query
    return `${filterResource} ${filterOperation} ${filterValue}`;
  } //if filter operation doesn't exist, default to Contains() method
  else return `Contains(${filterName},'${filterValue}')`;
};
```
My filter sources with suffixes look like this: quantity_ge, date_ge, date_le, partnumber_eq_str, and so on. For string fields where I'm filtering for a substring, I don't append a suffix and it just defaults to the Contains() method (source="partnumber" will filter for partnumber column using Contains()). Not sure if this type of solution is extensible enough for others, but thought I'd share some testing details

**Comment by jvert** [comment]
I agree that we need this feature and this is a clever way to do it. But I think using the _ separator is pretty fragile as it would break if your resource or value contained a _ character. I think that's not uncommon, e.g. "user_name" or something like that.

It's unfortunate that RA doesn't have a richer defined set of filters. I have hit problems before with just trying to express the difference between Contains and Equals. Maybe better approach is to smuggle odata filters in a different params value.

**Comment by 25747** [comment]
Yes I totally agree; to be a more universal solution it'd need to be some separator that couldn't appear in an SQL column or table name by definition. I did think about putting something more custom in the filter state or the URL params, but I needed a quick solution without getting into more custom filters in RA first. I will keep playing around with this over the next few days/weeks

**Comment by 25747** [comment]
> Yes I totally agree; to be a more universal solution it'd need to be some separator that couldn't appear in an SQL column or table name by definition. I did think about putting something more custom in the filter state or the URL params, but I needed a quick solution without getting into more custom filters in RA first. I will keep playing around with this over the next few days/weeks

The other quick thought I had earlier today was to pass the desired separator in along with the API URL and the options callback when setting up the dataprovider. This would allow everyone to choose their own separator/delimiter, and it keeps the ability to use the default RA filter properties "as intended".

**Comment by 25747** [comment]
> It's unfortunate that RA doesn't have a richer defined set of filters. I have hit problems before with just trying to express the difference between Contains and Equals. Maybe better approach is to smuggle odata filters in a different params value.

Sorry for triple-posting, I have a habit of thinking "out loud", but this will be my last comment for now; I was going through the RA 4.0 wish list and saw this from Francois: https://github.com/marmelab/react-admin/issues/5933#issuecomment-829428150

So this is clearly a react-admin limitation they are thinking about as well! My own little modification is working fine for me for the moment, so I am happy with what I have. Please feel free to leave this issue open or close it as you wish. I'd like to help contribute so let me know if/when you think you have some workable ideas! 
Cheers, 
Mike

**Comment by jvert** [comment]
fixed by #16

**Comment by te-online** [comment]
Hey, I'm trying to build a nested filter and can't quite figure out the best approach to doing it. Maybe @jvert you can help? 😊 

The resulting filter should look like `$filter=Items/any(c: c/SomeNumber eq 'filterValue')`. I would probably need to overwrite the request or something like that, right? Or is there a way to implement it using the logic in the mentioned PR?
